### PR TITLE
Part of pathfinding rework

### PR DESCRIPTION
**Pull request by LordOfLunacy**
I have picked up the task of preparing a rework of the pathfinding to allow for the addition of a new pathfinding implementation to the game. This initial pull request makes GuestPathfinding more modular by creating a GuestPathfinding base class and implementing the current pathfinding system into an OriginalPathfinding subclass.

What I would primarily like feedback on (other feedback is obviously welcome):
 - Where the GuestPathfinding object should be created (currently it's being done during initialization of GameState)
 - Suggestions of better ways to implement the pathfinding object that doesn't require a global variable reference

### CODE REVIEW

**Comment by LordOfLunacy** [comment]
I'm going to go ahead and mark this ready for review because my next plan of action is to make the goal position, queue ride index and ignore foreign queues into a part of the peep struct instead of being global variables which should probably be reviewed separately

File: src/openrct2/GameState.cpp
Comment (IntelOrca) [inline src/openrct2/GameState.cpp:93]: Remove, create with initialiser instead.

File: src/openrct2/entity/Staff.cpp
Comment (IntelOrca) [inline src/openrct2/entity/Staff.cpp:69]:
```suggestion
std::unique_ptr<GuestPathfinding> gGuestPathfinder = std::make_unique<GuestPathfinding>();
```

File: src/openrct2/peep/GuestPathfinding.h
Comment (IntelOrca) [inline src/openrct2/peep/GuestPathfinding.h:47]: This is a good time to improve the signatures. Use references instead of pointers when `nullptr` is not valid. And move to the TitleCase naming.

```suggestion
    virtual Direction ChooseDirection(const TileCoordsXYZ& loc, Peep& peep) = 0;
```

File: src/openrct2/peep/GuestPathfinding.h
Comment (IntelOrca) [inline src/openrct2/peep/GuestPathfinding.h:57]:
```suggestion
    virtual int32_t CalculateNextDestination(Guest& peep) = 0;
```

Also, use javadoc format, e.g.:
```cpp
/**
 * Overall guest pathfinding AI. Sets up Peep::DestinationX/DestinationY (which they move to in a
 * straight line, no pathfinding). Called whenever the guest has arrived at their previously set destination.
 *
 * @returns 0 if the guest has successfully had a new destination set up, nonzero otherwise.
 */
```

File: src/openrct2/peep/GuestPathfinding.h
Comment (IntelOrca) [inline src/openrct2/peep/GuestPathfinding.h:72]: These methods could change to use `Peep&`.

File: src/openrct2/peep/GuestPathfinding.h
Comment (IntelOrca) [inline src/openrct2/peep/GuestPathfinding.h:76]: What kind of better solution were you thinking of?

**Review by IntelOrca — CHANGES_REQUESTED** [review]

File: src/openrct2/peep/GuestPathfinding.h
Comment (LordOfLunacy) [reply src/openrct2/peep/GuestPathfinding.h:76]: I don't know really tbh, this code is so reliant on global variables though, it would probably be a lot of work to do anything else.

**Comment by ZehMatt** [comment]
In my opinion this is the wrong way to do it, the first thing that should be refactored is separating the behavior from path finding. Ideally the code would properly define a behavior tree for a guest and the staff and based on the next planned action it may or may not use path finding to build a route to walk on which then executes the action/behavior. I think abstraction is the wrong thing here, the behavior tree should decide what to do, the purpose of path finding is just selecting a route from A to B, no idea why you would want multiple models do that.

This is just my opinion

**Comment by LordOfLunacy** [comment]
> In my opinion this is the wrong way to do it, the first thing that should be refactored is separating the behavior from path finding. Ideally the code would properly define a behavior tree for a guest and the staff and based on the next planned action it may or may not use path finding to build a route to walk on which then executes the action/behavior. I think abstraction is the wrong thing here, the behavior tree should decide what to do, the purpose of path finding is just selecting a route from A to B, no idea why you would want multiple models do that.
> 
> This is just my opinion

Good point, the pathfinding itself does seem heavily embroiled with behavior logic where it doesn't need to be. Separating that out might be something that should be incorporated into this project.

 However, the pathfinding itself is a problem of its own due to its naive simplicity which limits its capability to properly identify routes and search for things (eg. current pathfinding is unable to take into account transportation, and will frequently identify wide paths as junctions). Abstracting the pathfinding allows us to give the option to keep the "vanilla" experience of the game, but implement a more advanced pathfinding system that is better suited to the task.

**Comment by Broxzier** [comment]
> eg. current pathfinding is unable to take into account transportation rides, and will frequently identify wide paths as junctions

The current pathfinding works for peeps because they use the same logic. If you'd change the way junctions are seen by the pathfinding, you will need to find a way to make guest walk towards the junctions. How would peeps know where to go when they walk on a tile that allows them to go multiple directions, but is not a junction itself? The junctions that it currently sees are essentially waypoint nodes.

**Comment by ZehMatt** [comment]
> > In my opinion this is the wrong way to do it, the first thing that should be refactored is separating the behavior from path finding. Ideally the code would properly define a behavior tree for a guest and the staff and based on the next planned action it may or may not use path finding to build a route to walk on which then executes the action/behavior. I think abstraction is the wrong thing here, the behavior tree should decide what to do, the purpose of path finding is just selecting a route from A to B, no idea why you would want multiple models do that.
> > This is just my opinion
> 
> Good point, the pathfinding itself does seem heavily embroiled with behavior logic where it doesn't need to be. Separating that out might be something that should be incorporated into this project.
> 
> However, the pathfinding itself is a problem of its own due to its naive simplicity which limits its capability to properly identify routes and search for things (eg. current pathfinding is unable to take into account transportation rides, and will frequently identify wide paths as junctions). Abstracting the pathfinding allows us to give the option to keep the "vanilla" experience of the game, but implement a more advanced pathfinding system that is better suited to the task.

I would suggest to try to build the path finding first independent of peeps. You want to have a system that returns a list of points/nodes for the Peep to walk towards. The peep has to just approach the nodes/points once the path finding query completed.

As for the behavior I would suggest to simply build the behavior tree without having any code, purely as the tree/graph, so one would be for Guests, one for each Staff type. Once that data is complete you can start implementing a system that uses that data to process the actions. This would also help de-hardcode some of the Guest logic as you can create more than one tree for the Guests, one could have the behavior for the peep to simply collect shop items and another one could be the peep just wanting to ride things until he vomits and so on, the imagination would be the limit here.

Its best to write those things independent of the current code, its better to have this stuff on-top first before replacing legacy code. Another benefit is that writing tests is quite easy since the code wouldn't rely on any other code or state then. You could also expose the new path finding to the scripting system and visualize/test it by building a plugin to like draw some lines or re-colour the tiles to see if you get the desired results.

You also mentioned transport rides, as for path finding this could be simply an option to include connections from stations or not, the peep has deal with going on transport rides as in, waiting for the queue, waiting for the ride to stop at the right station, exit and continue to navigate the planned route, but that kind of logic is not the job of the path finding algorithm besides returning a list of connections/nodes/pointers for traversal.

**Comment by LordOfLunacy** [comment]
> The current pathfinding works for peeps because they use the same logic. If you'd change the way junctions are seen by the pathfinding, you will need to find a way to make guest walk towards the junctions. How would peeps know where to go when they walk on a tile that allows them to go multiple directions, but is not a junction itself? The junctions that it currently sees are essentially waypoint nodes.

There is certainly more than 1 way to define a waypoint node, however, and the guest will actually refer to the choose direction function in the pathfinding API for the direction it should walk in.

 For example, you could feasibly create a junction node that's more than 1x1 in size that the guest has to walk towards, and once its in the junction, the guest pathfinding API would be able to give the guest directions to take a path to leave said junction through the correct exit to get to its next waypoint.

You could also have it set up so that you have a waypoint cluster (multiple 1x1 nodes make up a single junction, and the guest is navigating to a single waypoint to enter said cluster), so the guest makes it to a cluster entrance point, and based on the next cluster it wants to get to, the API is then able to guide the guest to the next cluster it so desires.

Within the current terms of the system, there are multiple pathfinding systems that could be deemed feasible with this GuestPathfinding class.

TL;DR junctions are abstract in regard to this API and the peeps' logic is equally unaware of it. All the peep sees is what direction to walk in next.

**Comment by ZehMatt** [comment]
> > > In my opinion this is the wrong way to do it, the first thing that should be refactored is separating the behavior from path finding. Ideally the code would properly define a behavior tree for a guest and the staff and based on the next planned action it may or may not use path finding to build a route to walk on which then executes the action/behavior. I think abstraction is the wrong thing here, the behavior tree should decide what to do, the purpose of path finding is just selecting a route from A to B, no idea why you would want multiple models do that.
> > > This is just my opinion
> > 
> > 
> > Good point, the pathfinding itself does seem heavily embroiled with behavior logic where it doesn't need to be. Separating that out might be something that should be incorporated into this project.
> > However, the pathfinding itself is a problem of its own due to its naive simplicity which limits its capability to properly identify routes and search for things (eg. current pathfinding is unable to take into account transportation rides, and will frequently identify wide paths as junctions). Abstracting the pathfinding allows us to give the option to keep the "vanilla" experience of the game, but implement a more advanced pathfinding system that is better suited to the task.
> 
> I would suggest to try to build the path finding first independent of peeps. You want to have a system that returns a list of points/nodes for the Peep to walk towards. The peep has to just approach the nodes/points once the path finding query completed.
> 
> As for the behavior I would suggest to simply build the behavior tree without having any code, purely as the tree/graph, so one would be for Guests, one for each Staff type. Once that data is complete you can start implementing a system that uses that data to process the actions. This would also help de-hardcode some of the Guest logic as you can create more than one tree for the Guests, one could have the behavior for the peep to simply collect shop items and another one could be the peep just wanting to ride things until he vomits and so on, the imagination would be the limit here.
> 
> Its best to write those things independent of the current code, its better to have this stuff on-top first before replacing legacy code. Another benefit is that writing tests is quite easy since the code wouldn't rely on any other code or state then. You could also expose the new path finding to the scripting system and visualize/test it by building a plugin to like draw some lines or re-colour the tiles to see if you get the desired results.
> 
> You also mentioned transport rides, as for path finding this could be simply an option to include connections from stations or not, the peep has deal with going on transport rides as in, waiting for the queue, waiting for the ride to stop at the right station, exit and continue to navigate the planned route, but that kind of logic is not the job of the path finding algorithm besides returning a list of connections/nodes/pointers for traversal.

A library I can recommend for FSM/Behavior is https://www.behaviortree.dev/ , once the tree is defined you will also better understand how Guests actually work currently.

**Comment by LordOfLunacy** [comment]
> > > In my opinion this is the wrong way to do it, the first thing that should be refactored is separating the behavior from path finding. Ideally the code would properly define a behavior tree for a guest and the staff and based on the next planned action it may or may not use path finding to build a route to walk on which then executes the action/behavior. I think abstraction is the wrong thing here, the behavior tree should decide what to do, the purpose of path finding is just selecting a route from A to B, no idea why you would want multiple models do that.
> > > This is just my opinion
> > 
> > 
> > Good point, the pathfinding itself does seem heavily embroiled with behavior logic where it doesn't need to be. Separating that out might be something that should be incorporated into this project.
> > However, the pathfinding itself is a problem of its own due to its naive simplicity which limits its capability to properly identify routes and search for things (eg. current pathfinding is unable to take into account transportation rides, and will frequently identify wide paths as junctions). Abstracting the pathfinding allows us to give the option to keep the "vanilla" experience of the game, but implement a more advanced pathfinding system that is better suited to the task.
> 
> I would suggest to try to build the path finding first independent of peeps. You want to have a system that returns a list of points/nodes for the Peep to walk towards. The peep has to just approach the nodes/points once the path finding query completed.
> 
> As for the behavior I would suggest to simply build the behavior tree without having any code, purely as the tree/graph, so one would be for Guests, one for each Staff type. Once that data is complete you can start implementing a system that uses that data to process the actions. This would also help de-hardcode some of the Guest logic as you can create more than one tree for the Guests, one could have the behavior for the peep to simply collect shop items and another one could be the peep just wanting to ride things until he vomits and so on, the imagination would be the limit here.
> 
> Its best to write those things independent of the current code, its better to have this stuff on-top first before replacing legacy code. Another benefit is that writing tests is quite easy since the code wouldn't rely on any other code or state then. You could also expose the new path finding to the scripting system and visualize/test it by building a plugin to like draw some lines or re-colour the tiles to see if you get the desired results.
> 
> You also mentioned transport rides, as for path finding this could be simply an option to include connections from stations or not, the peep has deal with going on transport rides as in, waiting for the queue, waiting for the ride to stop at the right station, exit and continue to navigate the planned route, but that kind of logic is not the job of the path finding algorithm besides returning a list of connections/nodes/pointers for traversal.

So to clarify, to start with you want to move the functions IsValidPathZandDirection and ChooseDirection out of the pathfinding entirely, and also move the logic for the guest to choose its next destination be isolated from the pathfinding.

 I do foresee a single issue with that and that's that, some of the logic like heading down dead-ends being more likely when the guest has food, would either need to be handled as a special case by the pathfinding API or the peep needs some of its own pathfinding capabilities as part of its own behavioral logic. 
 
 That could be dealt with, but the scale and scope of this change is on an order of magnitude difference from what is currently proposed, I definitely think that I want input from maintainers on what should be pursued here, simply because of the difference in scale and scope.

**Comment by janisozaur** [comment]
As part of this work, would it be possible to add some unit tests for pathfinding?

**Comment by LordOfLunacy** [comment]
> As part of this work (not necessarily this PR), would it be possible to add some unit tests for pathfinding?

That is part of the plan, after all, the current unit tests are fairly inadequate overall.

**Comment by LordOfLunacy** [comment]
This latest commit I believe resolves all of the requested changes

**Comment by LordOfLunacy** [comment]
I'm looking at it, and it appears that the only instance where gPeepPathFindGoalPosition, gPeepPathFindIgnoreForeignQueues or gPeepPathFindQueueRideIndex is called is changed outside of GuestPathfinding is when a mechanic is heading to inspection or answering a call to a broken down ride

**Comment by IntelOrca** [comment]
@LordOfLunacy good idea, but it is fine to leave for another PR.

File: src/openrct2/peep/GuestPathfinding.h
Comment (LordOfLunacy) [reply src/openrct2/peep/GuestPathfinding.h]: Your recommendation to change that to a unique_ptr solved the majority of my concerns with leaving that as a global.

Comment (IntelOrca) [reply src/openrct2/peep/GuestPathfinding.h]: Just becareful as you can not control the order of global unique_ptr destruction. So sometimes globals get disposed before other globals still use them during their dispose code.

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**
